share/getters/ipld.go

```diff
@@ -0,0 +1,126 @@
+)
+
+var _ share.Getter = (*IPLDGetter)(nil)
+
+// IPLDGetter is a share.Getter that retrieves shares from the IPLD network. An EDS store can be
+// provided to store shares retrieved from GetShares. Otherwise, result caching is handled by the
+// provided blockservice.
+type IPLDGetter struct {
+	rtrv  *eds.Retriever
+	bServ blockservice.BlockService
+
+	// store is an optional eds store that can be used to cache retrieved shares.
+	store *eds.Store
+}
+
+// NewIPLDGetter creates a new share.Getter that retrieves shares from the IPLD network.
+func NewIPLDGetter(bServ blockservice.BlockService) *IPLDGetter {
+	return NewIPLDGetterWithStore(bServ, nil)
+}
+
+// NewIPLDGetterWithStore creates a new IPLDGetter with an EDS store for storing EDSes from
+// GetShares.
+func NewIPLDGetterWithStore(bServ blockservice.BlockService, store *eds.Store) *IPLDGetter {
+	return &IPLDGetter{
+		rtrv:  eds.NewRetriever(bServ),
+		bServ: bServ,
+		store: store,
+	}
+}
+
+func (ig *IPLDGetter) GetShare(ctx context.Context, dah *share.Root, row, col int) (share.Share, error) {
+	root, leaf := ipld.Translate(dah, row, col)
+	nd, err := share.GetShare(ctx, ig.bServ, root, leaf, len(dah.RowsRoots))
+	if err != nil {
+		return nil, err
+	}
+
+	return nd, nil
+}
+
+func (ig *IPLDGetter) GetShares(ctx context.Context, root *share.Root) ([][]share.Share, error) {
+	// rtrv.Retrieve calls shares.GetShares until enough shares are retrieved to reconstruct the EDS
+	eds, err := ig.rtrv.Retrieve(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+
+	// store the EDS if a store was provided
+	if ig.store != nil {
+		err = ig.store.Put(ctx, root.Hash(), eds)
+		if err != nil {
+			return nil, fmt.Errorf("failed to cache retrieved shares: %w", err)
+		}
+	}
+
+	origWidth := int(eds.Width() / 2)
+	shares := make([][]share.Share, origWidth)
+
+	for i := 0; i < origWidth; i++ {
+		row := eds.Row(uint(i))
+		shares[i] = make([]share.Share, origWidth)
+		for j := 0; j < origWidth; j++ {
+			shares[i][j] = row[j]
+		}
+	}
+
+	return shares, nil
+}
+
+func (ig *IPLDGetter) GetSharesByNamespace(
+	ctx context.Context,
+	root *share.Root,
+	nID namespace.ID,
+) (share.NamespaceShares, error) {
+	if len(nID) != share.NamespaceSize {
+		return nil, fmt.Errorf("expected namespace ID of size %d, got %d", share.NamespaceSize, len(nID))
+	}
+
+	rowRootCIDs := make([]cid.Cid, 0)
+	for _, row := range root.RowsRoots {
+		if !nID.Less(nmt.MinNamespace(row, nID.Size())) && nID.LessOrEqual(nmt.MaxNamespace(row, nID.Size())) {
+			rowRootCIDs = append(rowRootCIDs, ipld.MustCidFromNamespacedSha256(row))
+		}
+	}
+	if len(rowRootCIDs) == 0 {
+		return nil, nil
+	}
+
+	errGroup, ctx := errgroup.WithContext(ctx)
+	shares := make([]share.RowNamespaceShares, len(rowRootCIDs))
+	for i, rootCID := range rowRootCIDs {
+		// shadow loop variables, to ensure correct values are captured
+		i, rootCID := i, rootCID
+		errGroup.Go(func() (err error) {
+			var proof *ipld.Proof
```

Comment:
This should not be nil so that we actually collect proofs. 
Should we do some basic tests to ensure we don't miss anything like this?

Reply:
Why is it nil? We collect proofs in GetSharesByNamespace! Yes we should add tests in this PR for it

Reply:
Vlad helped me understand here, disregard last comment

Reply:
Let's not disregard the testing part of your message))